src/scopeck.rs

```diff
@@ -151,156 +151,148 @@ struct ScopeState<'a> {
     frames_out: Vec<Frame>,
 }
 
-fn push_diagnostic(state: &mut ScopeState, ix: StatementIndex, diag: Diagnostic) {
-    state.diagnostics.entry(ix).or_insert(Vec::new()).push(diag);
-}
+impl<'a> ScopeState<'a> {
+    fn push_diagnostic(&mut self, ix: StatementIndex, diag: Diagnostic) {
```

Comment:
ScopeState certainly represents something: the scope pass state object. In a more object oriented approach like mmj2's, this would probably be called the Scoper or something, but it would be combined with the API for querying the data after the pass is done. In the current architecture, the state is a mutable object which provides methods to evolve it through the pass. This is consistent with the parameter list for these methods, which always has mutable ScopeState and immutable other parameters.

ScopeState is private (i.e. internal to scopeck module) but that doesn't mean it can't act in an object oriented fashion w.r.t the state evolution methods. Also it may be my Java upbringing but I prefer OO methods over global functions if there is a reasonable "self" object involved.

What does this proposed `-b` button do?